aiodataloader.py

```diff
@@ -1,29 +1,82 @@
-from asyncio import gather, ensure_future, get_event_loop, iscoroutine, iscoroutinefunction
-from collections import namedtuple
-from collections.abc import Iterable
+from asyncio import (
+    AbstractEventLoop,
+    Future,
+    gather,
+    ensure_future,
+    get_event_loop,
+    iscoroutine,
+    iscoroutinefunction,
+)
 from functools import partial
 
-from typing import List  # flake8: noqa
+from typing import (
+    Any,
+    Callable,
+    Coroutine,
+    Dict,
+    Generic,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    TypeVar,
+    Union,
+)
+from typing_extensions import Protocol, TypeGuard
+
+__version__ = "0.2.1"
+
+
+def iscoroutinefunctionorpartial(fn: Callable) -> TypeGuard[Callable[..., Coroutine]]:
+    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
 
-__version__ = '0.2.1'
+KeyT = TypeVar("KeyT")
+ReturnT = TypeVar("ReturnT")
+CacheKeyT = TypeVar("CacheKeyT")
+T = TypeVar("T")
 
-Loader = namedtuple('Loader', 'key,future')
 
+class BatchLoadFnProto(Protocol[KeyT, ReturnT]):
+    async def __call__(self, keys: List[KeyT]) -> List[ReturnT]:
+        ...
 
-def iscoroutinefunctionorpartial(fn):
-    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
+class Loader(tuple, Generic[KeyT, ReturnT]):
+    # NamedTuple has a metaclass conflict with Generic that prevents
+    # it from being used generically, and dataclasses aren't supported until
+    # Python 3.7
 
-class DataLoader(object):
+    key: KeyT
+    future: "Future[ReturnT]"
```

Comment:
This doesn't need to be quoted, does it?  That goes for most of the type hints added here.

Reply:
It does, because `Future` isn't subscriptable. This is still more specific than using `typing.Awaitable`, given how the library works (and it accepts the type parameters just fine)

![image](https://user-images.githubusercontent.com/57950/193991054-bc42493e-9f81-4620-b01c-6f18860201f0.png)

Reply:
Got it, I didn't notice because `Future` is subscriptable as of Python 3.9.  I would remove quoting on type hints in other cases unless it is necessary.  Or do you think quoting is generally preferred for some reason?

Reply:
In general I try to only quote when it's necessary (for example, inside DataLoader methods, the DataLoader self-reference needs to be quoted since at runtime it doesn't exist when the method is defined (when it isn't already DataLoaderT, but in that case DataLoaderT's bound argument gets quoted))